allspice/utils/bom_generation.py

```diff
@@ -248,31 +248,39 @@ def _extract_all_schdoc_components(
     """
 
     files_in_repo = repository.get_git_content(ref=ref)
-    all_components = []
-    for repo_file in files_in_repo:
-        if repo_file.name in schdoc_files_in_proj:
-            # If the file is not yet generated, we'll retry a few times.
-            retry_count = 0
-            while True:
-                retry_count += 1
-
-                try:
-                    schdoc_file_json = repository.get_generated_json(repo_file, ref=ref)
-                    all_components.extend(
-                        _extract_components_from_schdoc(
-                            schdoc_file_json,
-                            attributes_mapping,
-                        )
-                    )
 
-                    break
-                except NotYetGeneratedException:
-                    if retry_count > 20:
-                        break
+    schdoc_files_in_repo = [file for file in files_in_repo if file.name in schdoc_files_in_proj]
+
+    # Create a retry value for each file so that we can traverse them
+    # in parallel for improved performance
```

Comment:
This comment is a little misleading. This code doesn't do anything in parallel. I would also use a pretty different approach. We basically want to reimplement the JS client code or something even smarter. But I don't know if we care. So I guess this is fine.

Reply:
Hmmm, maybe I should look at our rate-limiting a little closer. My goal here was to limit any requests by this function to 4 per second. I didn’t implement something with promises/threads because I figured that could still potentially fire off dozens of requests simultaneously, and rate-limit us, even if we waited between requests for each endpoint.

Reply:
Our rate limiting session should already limit requests to X per second. If it isn't limiting enough, then we should probably adjust the `max_calls` and `period` arguments. The point of having a separate rate limiter is to pull out that concern so that rate limiting logic isn't sprinkled all over the codebase.

https://github.com/AllSpiceIO/py-allspice/blob/7fa769204240df1fa3692785d4554bbf9632b52e/allspice/ratelimiter.py#L5-L16

If you're interested, I can also post my pseudocode first draft of something that's more similar to the JS client code. It doesn't use promises or threads; that would be one of the improvements we could do.

Reply:
yeah, if you could share that it would be great

Reply:
See #58.